compiler/noirc_frontend/src/node_interner.rs

```diff
@@ -2283,20 +2304,28 @@ impl Methods {
 
                             if object.try_unify(typ, &mut bindings).is_ok() {
                                 Type::apply_type_bindings(bindings);
-                                return Some(method);
+                                candidate = Some(method);
+                                break;
                             }
                         }
                     } else {
-                        // Just return the first method whose name matches since we
-                        // can't match object types on static methods.
-                        return Some(method);
+                        candidate = Some(method);
+
+                        // If we recorded the concrete type this trait impl method belongs to,
+                        // and it matches typ, it's an exact match and we return that.
+                        if let Some(method_type) = method_type {
+                            if typ == method_type {
+                                break;
+                            }
+                        }
```

Comment:
Always avoid raw type equality! You should prefer `unify` or `try_unify` since only those will handle type variables and arithmetic generics correctly. In this case we should do the same we do on line 2305 where we check if try_unify is successful before applying the bindings and returning.

Also - why change this to use `candidate` instead of an early return? Looks like it requires more unnecessary code to me.

Reply:
This check is only done for primitive types but I think there may still be room for type variables, particularly ones bound to primitive types, to sneak in. It's also good practice to just avoid `==` for types in general IMO. It's just an unfortunate necessity to allow types in hashmaps in certain cases, although that does also mean you have to ensure there are no type variables in those types or be alright with some duplication.

Reply:
> Always avoid raw type equality!

I was thinking that if there's no self param we'll always have a concrete type... but I guess that if you have a `T` in some generic code and you do `T::method` that `T` could be a `FieldOrInteger`... not sure! In any case I'll change it.

> Also - why change this to use candidate instead of an early return?

I guess that makes sense. I was thinking that we don't always record the actual self type so it might not be correct to stop at the first match, but now I'm thinking that for a give type all of the methods will have a "self type" recorded, or all of them won't, so stopping at the first match is correct. I'll change it too.

Reply:
Any reason why `try_unify` is called on line 2305 instead of `unify`?

Reply:
> Any reason why try_unify is called on line 2305 instead of unify?

Hmm looks like it can be replaced. It was either a mistake or old enough from when `unify` unified type variables as it went even before it knew if the result succeeded. That should've been before we had `try_unify` at all though.